controllers/certificatesigningrequests_controller.go

```diff
@@ -19,45 +19,192 @@ package controllers
+	_ = r.log.WithValues("certificatesigningrequests", req.NamespacedName)
+
+	var err error
+	// Create a new signer using the private key that the instances will be configured with
+	r.signer, err = signer.Create(kubeTypes.NamespacedName{Namespace: r.watchNamespace,
+		Name: secrets.PrivateKeySecret}, r.client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "unable to create signer from private key secret")
+	}
+
+	certificateSigningRequest := &certificates.CertificateSigningRequest{}
+	if err := r.client.Get(ctx, req.NamespacedName, certificateSigningRequest); err != nil {
+		if k8sapierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	configMap, err := r.k8sclientset.CoreV1().ConfigMaps(r.watchNamespace).Get(ctx, InstanceConfigMap, meta.GetOptions{})
+	if err != nil {
+		if k8sapierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, r.reconcileCSR(certificateSigningRequest, configMap)
+}
+
+// reconcileCSR handles the CSR validation and approval
+func (r *certificateSigningRequestsReconciler) reconcileCSR(request *certificates.CertificateSigningRequest, configMap *core.ConfigMap) error {
+	if request == nil || configMap == nil {
+		return errors.New("CSR or instance ConfigMap cannot be nil")
+	}
 
-	// your logic here
+	// Get the list of instances that are expected to be Nodes
+	instances, err := parseHosts(configMap.Data)
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse hosts from ConfigMap %s", configMap.Name)
+	}
 
-	return ctrl.Result{}, nil
+	var instanceAddresses []string
+	for _, instance := range instances {
+		instanceAddresses = append(instanceAddresses, instance.Address)
+	}
+
+	// find the actual host names for all the instances present in configMap data
+	hostNames, err := r.findVMHostNames(instances)
+	if err != nil {
+		return errors.Wrapf(err, "unable to find host names from ConfigMap %s", configMap.Name)
+	}
+
+	certificateSigningRequest, err := csr.NewApprover(request, r.k8sclientset, instanceAddresses, hostNames, r.log)
+	if err != nil {
+		return errors.Wrapf(err, "could not create WMCO CSR Approver")
+	}
+
+	if err := certificateSigningRequest.Approve(); err != nil {
+		return errors.Wrapf(err, "WMCO CSR Approver could not approve %s CSR", request.Name)
+	}
```

Comment:
In the case we handle a CSR we do not want to approve, this will return error and the CSR will requeue indefinitely.

Reply:
no it will not return error if we do not want to approve CSR, it will only log in that case if validated = false(this is assuming you are talking about CSR's from non-BYOH instances).
It will only return error if there is error validating CSR due to parsing /the CSR contents are invalid(once it has been validated to be from an instance present in configmap)/ error with updating approval conditions.
[ValidateCSRContents() ](https://github.com/openshift/windows-machine-config-operator/pull/499/files#diff-666ed6f6b4dbed560da68743f3814f61952cd552103ea310e478cde0c5cbed2eR76) is something to look at to understand when the error is being returned.

I would think we need to requeue in the cases mentioned above as the node would not become available if we don't.